internals/overlord/state/notices.go

```diff
@@ -0,0 +1,378 @@
+	var jn jsonNotice
+	err := json.Unmarshal(data, &jn)
+	if err != nil {
+		return err
+	}
+	n.id = jn.ID
+	n.noticeType = NoticeType(jn.Type)
+	n.key = jn.Key
+	n.firstOccurred = jn.FirstOccurred
+	n.lastOccurred = jn.LastOccurred
+	n.lastRepeated = jn.LastRepeated
+	n.occurrences = jn.Occurrences
+	n.lastData = jn.LastData
+	if jn.RepeatAfter != "" {
+		n.repeatAfter, err = time.ParseDuration(jn.RepeatAfter)
+		if err != nil {
+			return fmt.Errorf("invalid repeat-after duration: %w", err)
+		}
+	}
+	if jn.ExpireAfter != "" {
+		n.expireAfter, err = time.ParseDuration(jn.ExpireAfter)
+		if err != nil {
+			return fmt.Errorf("invalid expire-after duration: %w", err)
+		}
+	}
+	return nil
+}
+
+type NoticeType string
+
+const (
+	// Recorded whenever a change is updated: when it is first spawned or its
+	// status was updated.
+	NoticeChangeUpdate NoticeType = "change-update"
+
+	// A client notice reported via the Pebble client API or "pebble notify".
+	// The key and data fields are provided by the user. The key must be in
+	// the format "mydomain.io/mykey" to ensure well-namespaced notice keys.
+	NoticeClient NoticeType = "client"
+
+	// Warnings are a subset of notices where the key is a human-readable
+	// warning message.
+	NoticeWarning NoticeType = "warning"
+)
+
+// NoticeTypeFromString validates the given string and returns the NoticeType,
+// or empty string if it's not valid.
+func NoticeTypeFromString(s string) NoticeType {
+	noticeType := NoticeType(s)
+	switch noticeType {
+	case NoticeChangeUpdate, NoticeClient, NoticeWarning:
+		return noticeType
+	default:
+		return ""
+	}
+}
+
+// AddNotice adds an occurrence of a notice with the specified type and key
+// and key-value data.
+func (s *State) AddNotice(noticeType NoticeType, key string, data map[string]string, repeatAfter time.Duration) {
+	if noticeType == "" || key == "" || len(key) > MaxNoticeKeyLength {
+		// Programming error
+		logger.Panicf("Internal error, please report: attempted to add invalid notice (type %q, key %q)",
+			noticeType, key)
+	}
+
+	s.writing()
+
+	now := time.Now().UTC()
+	uniqueKey := uniqueNoticeKey(noticeType, key)
+	notice, ok := s.notices[uniqueKey]
+	newOrRepeated := false
+	if !ok {
+		s.noticeId++
+		notice = &Notice{
+			id:            strconv.Itoa(s.noticeId),
+			noticeType:    noticeType,
+			key:           key,
+			firstOccurred: now,
+			lastRepeated:  now,
+			repeatAfter:   repeatAfter,
+			expireAfter:   7 * 24 * time.Hour, // one week
+			occurrences:   1,
+		}
+		s.notices[uniqueKey] = notice
+		newOrRepeated = true
+	} else {
+		notice.occurrences++
+		if repeatAfter != 0 && now.After(notice.lastRepeated.Add(repeatAfter)) {
+			// Update last repeated time if repeat-after time has elapsed
+			notice.lastRepeated = now
+			newOrRepeated = true
+		}
+	}
+	notice.lastOccurred = now
+	notice.lastData = data
+	notice.repeatAfter = repeatAfter
+
+	if newOrRepeated {
+		s.processNoticeWaiters()
```

Comment:
it feels a bit dangerous that this can potential block on unrelated code and timeouts decided by it, it would a bit more complex but safer to have in-between goroutines that distribute things to waiters

Reply:
The idea here is that `processNoticeWaiters` can't block, because it only sends to the current `noticeWaiters` channels, and those are added and removed with each call to `WaitNotice` (the other goroutines). And we have a done (`ctx.Done()`) channel on each send too, to avoid blocking if the context is cancelled.

The `addNoticeWaiter` / `removeNoticeWaiter` / `processNoticeWaiters` logic is intentionally very close together in that file, partly for that reason.

So I'm not sure how adding additional in-between goroutines will help here?